File: ios/loveaceios/ViewModel/AuthViewModel.swift
import Foundation
import os

private let logger = Logger(subsystem: "tech.loveace.loveaceios", category: "AuthViewModel")

enum AuthState { case initial, loading, reconnecting, authenticated, unauthenticated, error }

private enum ReauthenticationResult {
    case success(AUFEConnection)
    case unavailable
    case failed
}

@MainActor
@Observable
final class AuthViewModel {
    var state: AuthState = .initial
    var errorMessage: String?
    var userId: String = ""

    private let credentialStore = CredentialStore()
    private var heartbeatTask: Task<Void, Never>?
    private var authenticationTask: Task<Void, Never>?
    private var isReconnecting = false
    private var currentAuthenticationAttempt = UUID()

    private(set) var connection: AUFEConnection?
    private(set) var jwcService: JWCService?
    private(set) var yktService: YKTService?
    private(set) var isimService: ISIMService?
    private(set) var aacService: AACService?
    private(set) var laborClubService: LaborClubService?
    private(set) var competitionService: CompetitionService?
    private(set) var studentScheduleService: StudentScheduleService?
    private(set) var courseScheduleService: CourseScheduleService?
    private(set) var planService: PlanService?
    private(set) var repairService: RepairService?
    private(set) var teacherEvaluationService: TeacherEvaluationService?

    private(set) var sessionGeneration = 0

    var isAuthenticated: Bool { state == .authenticated }

    func login(userId: String, password: String, gatewayPassword: String? = nil) {
        authenticationTask?.cancel()
        stopHeartbeat()
        clearServices()
        let attempt = UUID()
        currentAuthenticationAttempt = attempt
        authenticationTask = Task {
            state = .loading; errorMessage = nil
            let ecPassword = gatewayPassword.flatMap { $0.isEmpty ? nil : $0 } ?? password
            let conn = AUFEConnection(userId: userId, ecPassword: ecPassword, password: password)
            await conn.startClient()

            let ecResult = await conn.ecLogin()
            guard !Task.isCancelled, attempt == currentAuthenticationAttempt else { return }
            guard ecResult.success else {
                let msg: String
                if ecResult.failInvalidCredentials { msg = "学号或密码错误" }
                else if ecResult.failMaybeAttacked { msg = "登录过于频繁，请稍后再试" }
                else if ecResult.failNetworkError { msg = "网络连接失败" }
                else { msg = "登录失败，请稍后重试" }
                Analytics.shared.trackLoginFailed(userId: userId, reason: msg)
                state = .error; errorMessage = msg; return
            }

            let uaapResult = await conn.uaapLogin()
            guard !Task.isCancelled, attempt == currentAuthenticationAttempt else { return }
            guard uaapResult.success else {
                let msg: String
                if uaapResult.failInvalidCredentials { msg = "学号或密码错误" }
                else if uaapResult.failNetworkError { msg = "网络连接失败" }
                else { msg = "教务系统登录失败" }
                Analytics.shared.trackLoginFailed(userId: userId, reason: msg)
                state = .error; errorMessage = msg; return
            }

            connection = conn
            initServices(conn)
            await wireSessionExpiredHandler(conn)
            startHeartbeat()
            credentialStore.save(UserCredentials(userId: userId, ecPassword: ecPassword, password: password))
            credentialStore.saveRemembered(UserCredentials(userId: userId, ecPassword: ecPassword, password: password))
            self.userId = userId
            Analytics.shared.trackLoginSuccess(userId: userId)
            state = .authenticated
            sessionGeneration += 1
            logger.info("Login successful")
        }
    }

    func restoreSession() {
        guard let creds = credentialStore.load() ?? credentialStore.loadRemembered() else {
            state = .unauthenticated; return
        }
        authenticationTask?.cancel()
        let attempt = UUID()
        currentAuthenticationAttempt = attempt
        authenticationTask = Task {
            state = .loading
            let conn = AUFEConnection(userId: creds.userId, ecPassword: creds.ecPassword, password: creds.password)
            await conn.startClient()
            let ec = await conn.ecLogin()
            guard !Task.isCancelled, attempt == currentAuthenticationAttempt else { return }
            guard ec.success else {
                Analytics.shared.trackLoginFailed(userId: creds.userId, reason: "restore_ec_failed")
                state = .unauthenticated; return
            }
            let uaap = await conn.uaapLogin()
            guard !Task.isCancelled, attempt == currentAuthenticationAttempt else { return }
            guard uaap.success else {
                Analytics.shared.trackLoginFailed(userId: creds.userId, reason: "restore_uaap_failed")
                state = .unauthenticated; return
            }
            connection = conn
            initServices(conn)
            await wireSessionExpiredHandler(conn)
            startHeartbeat()
            credentialStore.save(creds)
            userId = creds.userId
            Analytics.shared.trackLoginSuccess(userId: creds.userId)
            state = .authenticated
            sessionGeneration += 1
            logger.info("Session restored")
        }
    }

    func logout() {
        currentAuthenticationAttempt = UUID()
        authenticationTask?.cancel()
        authenticationTask = nil
        stopHeartbeat()
        clearServices()
        credentialStore.clear()
        Analytics.shared.clearUser()
        state = .unauthenticated
        logger.info("Logged out")
    }

    func verifyPassword(_ input: String) -> Bool {
        guard let creds = credentialStore.load() else { return false }
        return input == creds.password || input == creds.ecPassword
    }

    func getRememberedCredentials() -> UserCredentials? { credentialStore.loadRemembered() }
    func clearSavedCredentials() { credentialStore.clear(); credentialStore.clearRemembered() }

    private func startHeartbeat() {
        stopHeartbeat()
        heartbeatTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(240))
                guard let conn = connection, !Task.isCancelled else { break }
                switch await conn.heartbeat() {
                case .alive, .unavailable:
                    continue
                case .expired:
                    if !isReconnecting { await handleSessionExpired(for: conn) }
                }
            }
        }
    }

    private func stopHeartbeat() { heartbeatTask?.cancel(); heartbeatTask = nil }

    private func handleSessionExpired(for expiredConnection: AUFEConnection) async {
        guard !isReconnecting,
              state == .authenticated,
              let activeConnection = connection,
              activeConnection === expiredConnection else { return }
        isReconnecting = true
        defer { isReconnecting = false }
        state = .reconnecting
        logger.info("Auto-reconnecting...")
        Analytics.shared.trackSessionExpired(reason: "session_expired")

        guard let credentials = credentialStore.load() ?? credentialStore.loadRemembered() else {
            finishAutomaticSignOut()
            return
        }

        switch await reauthenticate(using: credentials) {
        case .success(let conn):
            connection = conn
            initServices(conn)
            await wireSessionExpiredHandler(conn)
            sessionGeneration += 1
            state = .authenticated
            startHeartbeat()
            Analytics.shared.trackSessionReconnectSuccess()
            logger.info("Auto-reconnect succeeded")
        case .unavailable:
            state = .authenticated
            startHeartbeat()
            logger.warning("Auto-reconnect deferred because the network is unavailable")
        case .failed:
            Analytics.shared.trackSessionReconnectFailed()
            finishAutomaticSignOut()
        }
    }

    private func reauthenticate(using credentials: UserCredentials) async -> ReauthenticationResult {
        let conn = AUFEConnection(
            userId: credentials.userId,
            ecPassword: credentials.ecPassword,
            password: credentials.password
        )
        await conn.startClient()
        let ec = await conn.ecLogin()
        if ec.failNetworkError { return .unavailable }
        guard ec.success else { return .failed }

        let uaap = await conn.uaapLogin()
        if uaap.failNetworkError { return .unavailable }
        guard uaap.success else { return .failed }
        return .success(conn)
    }

    private func finishAutomaticSignOut() {
        stopHeartbeat()
        clearServices()
        credentialStore.clear()
        currentAuthenticationAttempt = UUID()
        errorMessage = "会话已过期，请重新登录"
        state = .unauthenticated
    }

    private func wireSessionExpiredHandler(_ conn: AUFEConnection) async {
        await conn.setOnSessionExpired { [weak self, weak conn] in
            guard let conn else { return }
            Task { @MainActor [weak self] in
                await self?.handleSessionExpired(for: conn)
            }
        }
    }

    private func initServices(_ conn: AUFEConnection) {
        jwcService = JWCService(connection: conn)
        yktService = YKTService(connection: conn)
        isimService = ISIMService(connection: conn)
        aacService = AACService(connection: conn)
        laborClubService = LaborClubService(connection: conn)
        competitionService = CompetitionService(connection: conn)
        studentScheduleService = StudentScheduleService(connection: conn)
        courseScheduleService = CourseScheduleService(connection: conn)
        planService = PlanService(connection: conn)
        repairService = RepairService(connection: conn)
        teacherEvaluationService = TeacherEvaluationService(connection: conn)
    }

    private func clearServices() {
        connection = nil; jwcService = nil; yktService = nil; isimService = nil
        aacService = nil; laborClubService = nil; competitionService = nil
        studentScheduleService = nil; courseScheduleService = nil; planService = nil; repairService = nil
        teacherEvaluationService = nil
    }
}

File: android/app/src/main/java/tech/loveace/appv3/ui/viewmodel/AuthViewModel.kt
package tech.loveace.appv3.ui.viewmodel

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import tech.loveace.appv3.analytics.Analytics
import tech.loveace.appv3.data.local.CredentialStore
import tech.loveace.appv3.data.local.ScheduleStore
import tech.loveace.appv3.data.model.UserCredentials
import tech.loveace.appv3.data.network.AUFEConnection
import tech.loveace.appv3.data.service.*

enum class AuthState { Initial, Loading, Authenticated, Unauthenticated, Error }

data class AuthUiState(
    val state: AuthState = AuthState.Initial,
    val errorMessage: String? = null,
    val userId: String = "",
    val serviceGeneration: Long = 0,
)

/**
 * 认证 ViewModel
 * - 管理登录状态和所有服务实例
 * - 心跳保活：每 4 分钟 ping VPN，防止 session 过期
 * - Session 过期检测：HttpClient 嗅探到 VPN 登录页时自动重连
 */
class AuthViewModel(application: Application) : AndroidViewModel(application) {

    private val credentialStore = CredentialStore(application)

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()

    private var heartbeatJob: Job? = null
    private var isReconnecting = false
    private var serviceGeneration = 0L

    // Connection and services
    var connection: AUFEConnection? = null
        private set
    var jwcService: JWCService? = null
        private set
    var yktService: YKTService? = null
        private set
    var isimService: ISIMService? = null
        private set
    var aacService: AACService? = null
        private set
    var laborClubService: LaborClubService? = null
        private set
    var competitionService: CompetitionService? = null
        private set
    var studentScheduleService: StudentScheduleService? = null
        private set
    var courseScheduleService: CourseScheduleService? = null
        private set
    var planService: PlanService? = null
        private set
    var repairService: RepairService? = null
        private set
    var teacherEvaluationService: TeacherEvaluationService? = null
        private set

    val isAuthenticated get() = _uiState.value.state == AuthState.Authenticated

    // ==================== Login ====================

    fun login(userId: String, ecPassword: String, password: String) {
        viewModelScope.launch {
            _uiState.value = AuthUiState(state = AuthState.Loading)
            try {
                val conn = AUFEConnection(userId, ecPassword, password)
                conn.startClient()

                // EC (VPN) login
                val ecResult = conn.ecLogin()
                if (!ecResult.success) {
                    val msg = when {
                        ecResult.failInvalidCredentials -> "VPN 账号或密码错误"
                        ecResult.failMaybeAttacked -> "登录过于频繁，请稍后再试"
                        ecResult.failNetworkError -> "网络连接失败"
                        else -> "VPN 登录失败"
                    }
                    Analytics.trackLoginFailed(userId, msg)
                    _uiState.value = AuthUiState(state = AuthState.Error, errorMessage = msg)
                    return@launch
                }

                // UAAP (CAS) login
                val uaapResult = conn.uaapLogin()
                if (!uaapResult.success) {
                    val msg = when {
                        uaapResult.failInvalidCredentials -> "教务系统密码错误"
                        uaapResult.failNetworkError -> "网络连接失败"
                        else -> "UAAP 登录失败"
                    }
                    Analytics.trackLoginFailed(userId, msg)
                    _uiState.value = AuthUiState(state = AuthState.Error, errorMessage = msg)
                    return@launch
                }

                connection = conn
                initServices(conn)
                wireSessionExpiredHandler(conn)
                startHeartbeat()

                // 始终保存凭证（用于会话恢复和快速登录）
                credentialStore.save(UserCredentials(userId, ecPassword, password))
                credentialStore.saveRemembered(UserCredentials(userId, ecPassword, password))

                _uiState.value = AuthUiState(
                    state = AuthState.Authenticated,
                    userId = userId,
                    serviceGeneration = serviceGeneration,
                )
                Analytics.trackLoginSuccess(userId)
                Log.i(TAG, "✅ Login successful: $userId")
            } catch (e: Exception) {
                Log.e(TAG, "Login error", e)
                Analytics.trackLoginFailed(userId, "登录异常")
                _uiState.value = AuthUiState(state = AuthState.Error, errorMessage = "登录异常: ${e.message}")
            }
        }
    }

    // ==================== Session Restore ====================

    fun restoreSession() {
        // 优先用 session 凭证，没有则用 remembered 凭证（退出登录后快速登录场景）
        val creds = credentialStore.load() ?: credentialStore.loadRemembered() ?: run {
            _uiState.value = AuthUiState(state = AuthState.Unauthenticated)
            return
        }
        viewModelScope.launch {
            _uiState.value = AuthUiState(state = AuthState.Loading)
            try {
                val conn = AUFEConnection(creds.userId, creds.ecPassword, creds.password)
                conn.startClient()

                val ecResult = conn.ecLogin()
                if (!ecResult.success) {
                    Analytics.trackLoginFailed(creds.userId, "VPN 登录失败")
                    _uiState.value = AuthUiState(state = AuthState.Unauthenticated)
                    return@launch
                }

                val uaapResult = conn.uaapLogin()
                if (!uaapResult.success) {
                    Analytics.trackLoginFailed(creds.userId, "UAAP 登录失败")
                    _uiState.value = AuthUiState(state = AuthState.Unauthenticated)
                    return@launch
                }

                connection = conn
                initServices(conn)
                wireSessionExpiredHandler(conn)
                startHeartbeat()

                // 恢复 session 凭证
                credentialStore.save(creds)

                _uiState.value = AuthUiState(
                    state = AuthState.Authenticated,
                    userId = creds.userId,
                    serviceGeneration = serviceGeneration,
                )
                Analytics.trackLoginSuccess(creds.userId)
                Log.i(TAG, "✅ Session restored: ${creds.userId}")
            } catch (e: Exception) {
                Log.e(TAG, "Session restore failed", e)
                Analytics.trackLoginFailed(creds.userId, "会话恢复失败")
                _uiState.value = AuthUiState(state = AuthState.Unauthenticated)
            }
        }
    }

    // ==================== Logout ====================

    fun logout() {
        stopHeartbeat()
        viewModelScope.launch {
            connection?.close()
            clearServices()
            credentialStore.clear() // 只清 session 凭证，保留 remembered 用于快速登录
            Analytics.clearUser()
            _uiState.value = AuthUiState(state = AuthState.Unauthenticated)
            Log.i(TAG, "👋 Logged out")
        }
    }

    // ==================== Password Verification (for YKT) ====================

    fun verifyPassword(input: String): Boolean {
        val creds = credentialStore.load() ?: return false
        return input == creds.password || input == creds.ecPassword
    }

    fun getRememberedCredentials(): UserCredentials? = credentialStore.loadRemembered()

    fun clearSavedCredentials() {
        credentialStore.clear()
        credentialStore.clearRemembered()
        Analytics.clearUser()
    }

    // ==================== Heartbeat Keepalive ====================

    private fun startHeartbeat() {
        stopHeartbeat()
        heartbeatJob = viewModelScope.launch(Dispatchers.IO) {
            Log.i(TAG, "💓 Heartbeat started (interval: ${HEARTBEAT_INTERVAL_MS / 1000}s)")
            while (isActive) {
                delay(HEARTBEAT_INTERVAL_MS)
                val conn = connection ?: break
                val alive = conn.heartbeat()
                if (!alive && !isReconnecting) {
                    Log.w(TAG, "💓 Heartbeat detected session expired, reconnecting...")
                    handleSessionExpired()
                }
            }
        }
    }

    private fun stopHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = null
    }

    // ==================== Session Expired Handler ====================

    private fun wireSessionExpiredHandler(conn: AUFEConnection) {
        conn.setOnSessionExpired {
            if (!isReconnecting) {
                Log.w(TAG, "⚠️ HttpClient detected session expired")
                viewModelScope.launch { handleSessionExpired() }
            }
        }
    }

    private suspend fun handleSessionExpired() {
        if (isReconnecting) return
        isReconnecting = true
        Analytics.trackSessionExpired("session_expired")
        try {
            val conn = connection ?: return
            Log.i(TAG, "🔄 Auto-reconnecting...")
            val success = conn.reconnect()
            if (success) {
                // Re-wire callback (new clients after reconnect)
                wireSessionExpiredHandler(conn)
                // Re-init services with refreshed connection
                initServices(conn)
                _uiState.value = _uiState.value.copy(serviceGeneration = serviceGeneration)
                Analytics.trackSessionReconnectSuccess()
                Log.i(TAG, "✅ Auto-reconnect succeeded")
            } else {
                Log.e(TAG, "❌ Auto-reconnect failed, user needs to re-login")
                Analytics.trackSessionReconnectFailed()
                stopHeartbeat()
                _uiState.value = AuthUiState(
                    state = AuthState.Error,
                    errorMessage = "会话已过期，请重新登录"
                )
            }
        } finally {
            isReconnecting = false
        }
    }

    // ==================== Service Management ====================

    private fun initServices(conn: AUFEConnection) {
        jwcService = JWCService(conn)
        yktService = YKTService(conn)
        isimService = ISIMService(conn)
        aacService = AACService(conn)
        laborClubService = LaborClubService(conn)
        competitionService = CompetitionService(conn)
        studentScheduleService = StudentScheduleService(conn)
        courseScheduleService = CourseScheduleService(conn)
        planService = PlanService(conn)
        repairService = RepairService(conn)
        teacherEvaluationService = TeacherEvaluationService(conn)
        serviceGeneration++
    }

    private fun clearServices() {
        connection = null
        jwcService = null
        yktService = null
        isimService = null
        aacService = null
        laborClubService = null
        competitionService = null
        studentScheduleService = null
        courseScheduleService = null
        planService = null
        repairService = null
        teacherEvaluationService = null
    }

    override fun onCleared() {
        super.onCleared()
        stopHeartbeat()
        viewModelScope.launch { connection?.close() }
    }

    companion object {
        private const val TAG = "AuthViewModel"
        private const val HEARTBEAT_INTERVAL_MS = 4 * 60 * 1000L // 4 minutes
    }
}
